mypy/plugins/dataclasses.py

```diff
@@ -426,28 +433,46 @@ def collect_attributes(self) -> Optional[List[DataclassAttribute]]:
             # If we find any attribute that is_in_init, not kw_only, and that
             # doesn't have a default after one that does have one,
             # then that's an error.
-            if found_default and attr.is_in_init and not attr.has_default and not attr.kw_only:
-                # If the issue comes from merging different classes, report it
-                # at the class definition point.
-                context = (Context(line=attr.line, column=attr.column) if attr in attrs
-                           else ctx.cls)
-                ctx.api.fail(
-                    'Attributes without a default cannot follow attributes with one',
-                    context,
-                )
-
+            context = (
+                Context(line=attr.line, column=attr.column)
+                if attr in attrs
+                else ctx.cls
+            )
+            self._validate_field(
+                attr,
+                found_default=found_default,
+                found_kw_sentinel=found_kw_sentinel,
+                context=context,
+            )
             found_default = found_default or (attr.has_default and attr.is_in_init)
-            if found_kw_sentinel and self._is_kw_only_type(attr.type):
-                context = (Context(line=attr.line, column=attr.column) if attr in attrs
-                           else ctx.cls)
-                ctx.api.fail(
-                    'There may not be more than one field with the KW_ONLY type',
-                    context,
-                )
             found_kw_sentinel = found_kw_sentinel or self._is_kw_only_type(attr.type)
 
         return all_attrs
 
+    def _validate_field(
+        self,
+        attr: DataclassAttribute,
+        *,
+        found_default: bool,
+        found_kw_sentinel: bool,
+        context: Optional[Context] = None,
+    ) -> None:
+        if context is None:
+            context = self._ctx.cls
+        if found_default and attr.is_in_init and not attr.has_default and not attr.kw_only:
+            # If the issue comes from merging different classes, report it
```

Comment:
This comment would seem to fit better on line 461.

Reply:
I've just copy pasted this part 🙂